src/java.base/share/classes/java/lang/invoke/X-VarHandleSegmentView.java.template

```diff
@@ -39,532 +39,742 @@ import static java.lang.invoke.MethodHandleStatics.UNSAFE;
+#else[byte]
+        return (alignmentMask & NON_PLAIN_ACCESS_MIN_ALIGN_MASK) != NON_PLAIN_ACCESS_MIN_ALIGN_MASK ?
+                (fixedOffset ? Unaligned.FORM : Unaligned.OFFSET_FORM) :
+                (fixedOffset ? Aligned.FORM : Aligned.OFFSET_FORM);
+#end[byte]
+    }
 
-    static final int NON_PLAIN_ACCESS_MIN_ALIGN_MASK = $BoxType$.BYTES - 1;
+    static {#if[byte]?final:sealed} class Unaligned {
 
-    static final VarForm FORM = new VarForm(VarHandleSegmentAs$Type$s.class, MemorySegment.class, $type$.class, MemoryLayout.class, long.class, long.class);
+#if[!byte]
+        static final boolean BE = UNSAFE.isBigEndian();
 
-    VarHandleSegmentAs$Type$s(boolean be, long alignmentMask, boolean exact) {
-        super(FORM, be, alignmentMask, exact);
-    }
+#end[byte]
+        static final ScopedMemoryAccess SCOPED_MEMORY_ACCESS = ScopedMemoryAccess.getScopedMemoryAccess();
 
-    @Override
-    final MethodType accessModeTypeUncached(VarHandle.AccessType accessType) {
-        return accessType.accessModeType(MemorySegment.class, $type$.class, MemoryLayout.class, long.class, long.class);
-    }
+        static final VarForm FORM = new VarForm(Unaligned.class, MemorySegment.class, $type$.class, long.class);
+        static final VarForm OFFSET_FORM = new VarForm(Unaligned.class, MemorySegment.class, $type$.class, long.class, long.class);
 
-    @Override
-    public VarHandleSegmentAs$Type$s withInvokeExactBehavior() {
-        return hasInvokeExactBehavior() ?
-                this :
-                new VarHandleSegmentAs$Type$s(be, alignmentMask, true);
-    }
+        Unaligned() { throw new AssertionError(); }
 
-    @Override
-    public VarHandleSegmentAs$Type$s withInvokeBehavior() {
-        return !hasInvokeExactBehavior() ?
-                this :
-                new VarHandleSegmentAs$Type$s(be, alignmentMask, false);
-    }
+        @ForceInline
+        static long offset(AbstractMemorySegmentImpl bb, long base, long offset) {
+            long segment_base = bb.unsafeGetOffset();
+            return segment_base + base + offset;
+        }
 
-#if[floatingPoint]
-    @ForceInline
-    static $rawType$ convEndian(boolean big, $type$ v) {
-        $rawType$ rv = $Type$.$type$ToRaw$RawType$Bits(v);
-        return big == BE ? rv : $RawBoxType$.reverseBytes(rv);
-    }
+        @ForceInline
+        static AbstractMemorySegmentImpl checkSegment(Object obb, VarHandleSegmentViewBase handle, long base, boolean ro) {
+            AbstractMemorySegmentImpl oo = (AbstractMemorySegmentImpl)Objects.requireNonNull(obb);
+            oo.checkEnclosingLayout(base, handle.enclosing, ro);
+            return oo;
+        }
 
-    @ForceInline
-    static $type$ convEndian(boolean big, $rawType$ rv) {
-        rv = big == BE ? rv : $RawBoxType$.reverseBytes(rv);
-        return $Type$.$rawType$BitsTo$Type$(rv);
-    }
+        @ForceInline
+        static $type$ get(VarHandle ob, Object obb, long base) {
```

Comment:
I like this: a fast path for non-strided access, while, at the same time, reusing the impl.
In the future we can also add another variant for when there's only one index (which is by far the most common case).

Reply:
I think that would require new impl classes, as there may be invocation type clash with the existing one-offset version. And the new offset will require extra checks too.